include/lo2s/perf/reader.hpp

```diff
@@ -0,0 +1,368 @@
+                std::uint64_t config1 = 0, std::set<Cpu> cpus = std::set<Cpu>());
+
+    const std::set<Cpu>& supported_cpus() const;
+
+    // automatically determine availability
+    void set_availability();
+
+    const Availability& get_availability() const
+    {
+        return availability_;
+    };
+
+    bool is_valid() const;
+    bool is_available_in(ExecutionScope scope) const
+    {
+        // per-process should always work. the counter will just not count if the process is
+        // scheduled on a core that is not supprted by that counter
+        return scope.is_thread() || cpus_.empty() || cpus_.count(scope.as_cpu());
+    }
+
+    std::string get_name() const
+    {
+        return name_;
+    }
+
+    std::set<Cpu> get_cpus() const
+    {
+        return cpus_;
+    }
+
+    void set_common_attrs(std::string name, perf_type_id type, std::uint64_t config,
+                          std::uint64_t config1 = 0, std::set<Cpu> cpus = std::set<Cpu>())
+    {
+        name_ = name;
+        cpus_ = cpus;
+        attr_.type = type;
+        attr_.config = config;
+        attr_.config1 = config1;
+    }
+
+    void as_counter();
+    void as_group_leader();
+
+private:
+    std::string name_ = "";
+    std::set<Cpu> cpus_;
+    Availability availability_ = Availability::UNAVAILABLE;
+};
+
+/**
+ * Contains an event parsed from sysfs
+ * @note call on as_sample() after creation to get a valid
+ * event, otherwise the availability will be set to UNAVAILABLE
+ */
+class SysfsEvent : public PredefEvent
+{
+public:
+    using PredefEvent::PredefEvent;
+
+    void as_sample();
+    bool degrade_percision();
+};
+
+class TimeEvent : public PerfEvent
+{
+public:
+    TimeEvent(bool enable_on_exec, uint64_t addr);
+};
+
+/**
+ * Contains an event that is addressable via name
+ * @note Call on as_syscall() after creation if needed
+ */
+class TracepointEvent : public PerfEvent
+{
+public:
+    class ParseError : public std::runtime_error
+    {
+    public:
+        ParseError(const std::string& what) : std::runtime_error(what)
+        {
+        }
+
+        ParseError(const std::string& what, int error_code);
+    };
+
+    TracepointEvent() : id_(-1)
+    {
+    }
+    TracepointEvent(bool enable_on_exec, int event_id);
+    TracepointEvent(bool enable_on_exec, const std::string& name);
+
+    friend bool operator==(const TracepointEvent& lhs, const TracepointEvent& rhs)
+    {
+        return lhs.id_ == rhs.id_;
+    }
+
+    friend bool operator<(const TracepointEvent& lhs, const TracepointEvent& rhs)
+    {
+        return lhs.id_ < rhs.id_;
+    }
+
+    friend bool operator>(const TracepointEvent& lhs, const TracepointEvent& rhs)
+    {
+        return lhs.id_ > rhs.id_;
+    }
+
+    void parse_format();
+
+    void as_syscall()
+    {
+        attr_.sample_type |= PERF_SAMPLE_IDENTIFIER;
+    }
+
+    const auto& fields() const
+    {
+        return fields_;
+    }
+
+    const auto& common_fields() const
```

Comment:
The common_fields_ member also seems to not be used anywhere, delete that too?
Not sure, because that would mean to throw all fields that start with "common_" away.

Reply:
A quick test shows that we are not recording the fields starting with common_ anyways currently, so we  can remove it.